File: SkipList/SkipList.java
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class SkipList {
    private static final int MAX_LEVEL = 16;
    private Node header;
    private int level;

    public SkipList() {
        level = 0;
        header = new Node(Integer.MIN_VALUE, MAX_LEVEL);
    }

    private int randomLevel() {
        Random rand = new Random();
        int lvl = 0;
        while (lvl < MAX_LEVEL && rand.nextDouble() < 0.5)
            lvl++;
        return lvl;
    }

    public void insertElement(int key) {
        int level = randomLevel();
        Node[] update = new Node[level + 1];
        Node current = header;

        for (int i = level; i >= 0; i--) {
            while (current.forward.get(i) != null && current.forward.get(i).key < key)
                current = current.forward.get(i);
            update[i] = current;
        }
        current = current.forward.get(0);

        if (current == null || current.key != key) {
            Node newNode = new Node(key, level);
            for (int i = 0; i <= level; i++) {
                newNode.forward.set(i, update[i].forward.get(i));
                update[i].forward.set(i, newNode);
            }
        }
    }

    public void deleteElement(int key) {
        Node[] update = new Node[level + 1];
        Node current = header;

        for (int i = level; i >= 0; i--) {
            while (current.forward.get(i) != null && current.forward.get(i).key < key)
                current = current.forward.get(i);
            update[i] = current;
        }
        current = current.forward.get(0);

        if (current != null && current.key == key) {
            for (int i = 0; i < level; i++) {
                if (update[i].forward.get(i) != current)
                    break;
                update[i].forward.set(i, current.forward.get(i));
            }
        }
    }

    public void searchElement(int key) {
        Node current = header;
        for (int i = level; i >= 0; i--) {
            while (current.forward.get(i) != null && current.forward.get(i).key < key)
                current = current.forward.get(i);
        }
        current = current.forward.get(0);
        if (current != null && current.key == key)
            System.out.println("Found key: " + key);
        else
            System.out.println("Not Found key: " + key);
    }

    public void displayList() {
        System.out.println("Skip List:");
        for (int i = 0; i <= level; i++) {
            Node node = header.forward.get(i);
            System.out.print("Level " + i + ": ");
            while (node != null) {
                System.out.print(node.key + " ");
                node = node.forward.get(i);
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        SkipList skipList = new SkipList();

        // Insert elements
        skipList.insertElement(3);
        skipList.insertElement(6);
        skipList.insertElement(7);
        skipList.insertElement(9);
        skipList.insertElement(12);
        skipList.insertElement(19);
        skipList.insertElement(17);
        skipList.insertElement(26);
        skipList.insertElement(21);
        skipList.insertElement(25);

        // Display the skip list
        skipList.displayList();

        // Search for an element
        skipList.searchElement(19);

        // Delete an element
        skipList.deleteElement(19);
        skipList.displayList();
    }
}

class Node {
    int key;
    List<Node> forward;

    Node(int key, int level) {
        this.key = key;
        forward = new ArrayList<>(level + 1);
        for (int i = 0; i <= level; i++)
            forward.add(null);
    }
}

File: SkipList/skipList.py
import random

class Node:
    def __init__(self, key, level):
        self.key = key
        self.forward = [None] * (level + 1)

class SkipList:
    MAX_LEVEL = 16

    def __init__(self):
        self.header = Node(float("-inf"), self.MAX_LEVEL)
        self.level = 0

    def random_level(self):
        level = 0
        while level < self.MAX_LEVEL and random.random() < 0.5:
            level += 1
        return level

    def insert_element(self, key):
        level = self.random_level()
        update = [None] * (level + 1)
        current = self.header

        for i in range(level, -1, -1):
            while current.forward[i] and current.forward[i].key < key:
                current = current.forward[i]
            update[i] = current

        current = current.forward[0]

        if not current or current.key != key:
            new_node = Node(key, level)
            for i in range(level + 1):
                new_node.forward[i] = update[i].forward[i]
                update[i].forward[i] = new_node

    def delete_element(self, key):
        update = [None] * (self.level + 1)
        current = self.header

        for i in range(self.level, -1, -1):
            while current.forward[i] and current.forward[i].key < key:
                current = current.forward[i]
            update[i] = current

        current = current.forward[0]

        if current and current.key == key:
            for i in range(self.level + 1):
                if update[i].forward[i] != current:
                    break
                update[i].forward[i] = current.forward[i]

    def search_element(self, key):
        current = self.header
        for i in range(self.level, -1, -1):
            while current.forward[i] and current.forward[i].key < key:
                current = current.forward[i]
        current = current.forward[0]
        if current and current.key == key:
            print("Found key:", key)
        else:
            print("Not Found key:", key)

    def display_list(self):
        print("Skip List:")
        for i in range(self.level + 1):
            node = self.header.forward[i]
            print("Level", i, end=": ")
            while node:
                print(node.key, end=" ")
                node = node.forward[i]
            print()

if __name__ == "__main__":
    skip_list = SkipList()

    for i in range(0, 25):
        key = random.randrange(1000)
        skip_list.insert_element(key)
        skip_list.display_list()

    # Insert elements
    # skip_list.insert_element(3)
    # skip_list.insert_element(6)
    # skip_list.insert_element(7)
    # skip_list.insert_element(9)
    # skip_list.insert_element(12)
    # skip_list.insert_element(19)
    # skip_list.insert_element(17)
    # skip_list.insert_element(26)
    # skip_list.insert_element(21)
    # skip_list.insert_element(25)

    # # Display the skip list
    # skip_list.display_list()

    # # Search for an element
    # skip_list.search_element(19)

    # # Delete an element
    # skip_list.delete_element(19)
    # skip_list.display_list()